task_sdk/src/airflow/sdk/api/client.py

```diff
@@ -219,7 +219,24 @@ def get(
         params = {}
         if map_index is not None:
             params.update({"map_index": map_index})
-        resp = self.client.get(f"xcoms/{dag_id}/{run_id}/{task_id}/{key}", params=params)
+        try:
+            resp = self.client.get(f"xcoms/{dag_id}/{run_id}/{task_id}/{key}", params=params)
+        except ServerResponseError as e:
+            if e.response.status_code == HTTPStatus.NOT_FOUND:
+                log.error(
+                    "XCom not found",
+                    dag_id=dag_id,
+                    run_id=run_id,
+                    task_id=task_id,
+                    key=key,
+                    map_index=map_index,
+                    detail=e.detail,
+                    status_code=e.response.status_code,
+                )
+                # Airflow 2.x just ignores the absence of an XCom and moves on with a return value of None
+                # Hence returning with key as `key` and value as `None`, so that the message is sent back to task runner
+                # and the default value of None in xcom_pull is used.
+                return XComResponse(key=key, value=None)
```

Comment:
Should this re-raise the exception for other errors?

Reply:
Uuups, review problem. MUST re-raise other!

Reply:
We have an event hook defined here: https://github.com/apache/airflow/blob/main/task_sdk/src/airflow/sdk/api/client.py#L300
With handlers here: https://github.com/apache/airflow/blob/main/task_sdk/src/airflow/sdk/api/client.py#L92

The response hook makes sure that only a ServerResponseError is raised, if there is an error:
https://github.com/apache/airflow/blob/main/task_sdk/src/airflow/sdk/api/client.py#L356-L392

Reply:
No I mean with the current code, if a ServerResponseError that’s not 404 is raised by the XCom API, this would silently suppress that exception and fall over the the line after this. Since `resp` is not defined if the `try` block is unsuccessful, this would result in this function failing with a very unhelpful `name 'resp' is not defined`. A `raise` statement should be inserted after line 239 to retain the original exception.

Reply:
(And it mayb be a good idea to have a test case for non-404 exceptions)

Reply:
Okay, I misunderstood on the type of exception. Let me add that and add a test case, for another case

Reply:
I added a case for 500 (which is a valid response from the api) and tested it out by hacking the API to send that to see if the scheduler crashes, it doesn't all good.